tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -12688,16 +12684,12 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC, Expr *BitWidth,
       D.setInvalidType();
       return false;
     }
-    // Add methods if not ready.
-    hlslSource->AddHLSLObjectMethodsIfNotReady(qt);
   } else if (qt->isArrayType()) {
     QualType eltQt(qt->getArrayElementTypeNoTypeQual(), 0);
     while (eltQt->isArrayType())
       eltQt = QualType(eltQt->getArrayElementTypeNoTypeQual(), 0);
 
     if (hlsl::IsObjectType(this, eltQt, &bDeprecatedEffectObject)) {
-      // Add methods if not ready.
-      hlslSource->AddHLSLObjectMethodsIfNotReady(eltQt);
```

Comment:
Where do these get "completed" now? Looks like RequireCompleteType is called all over the place?

Reply:
This is the best part. Clang enforces C rules for requiring completeness. Any use of a type that requires completeness in C also requires it in HLSL. Whenever a type is required to be complete, but is not, clang calls out to the External AST source (if available) to complete the type. This preserves our laziness, but through clang’s built in methods.